README.md
# React Native In-App Browser

[![npm (scoped)][npm_shield]][npm] ![license: mit][license_shield]

In-App browser support for React Native, using [Chrome Custom Tabs][chromecustomtabs]
on Android
and [Safari View Controller][sfsafariviewcontroller] on iOS.

## Compatibility

⚠️**Support has ended on October 1st 2020.**

It is recommended migrating to [react-native-inappbrowser-reborn] for continued support and more features.

|   React Native    |                  Library                  |   Status   | End-of-Life |
| :---------------: | :---------------------------------------: | :--------: | :---------: |
| `0.60.0 - 0.62.x` |    [![npm (scoped)][npm_shield]][npm]     |  **EOL**   | 2020-10-01  |
| `0.57.x - 0.59.x` | [![npm v2 (scoped)][npm_2_shield]][npm_2] |  **EOL**   | 2020-01-03  |
| `0.57.x - 0.59.5` | [![npm v1 (scoped)][npm_1_shield]][npm_1] |  **EOL**   | 2019-06-09  |

**Using Expo?** Check out [WebBrowser](https://docs.expo.io/versions/latest/sdk/webbrowser/).

## Installation

Install the package via Yarn or npm:

```sh
yarn add @matt-block/react-native-in-app-browser

# or
npm install --save @matt-block/react-native-in-app-browser
```

## Linking

### React Native `0.60.0` and later

Packages are [autolinked](https://github.com/react-native-community/cli/blob/master/docs/autolinking.md),
however an extra step for iOS projects is needed:

```sh
cd ios && pod install && cd ..
```

### React Native `0.59.x` and below

Manually link the native module to your project:

```sh
react-native link @matt-block/react-native-in-app-browser
```

### Note for iOS projects

In order for Xcode projects to build when using Swift-based libraries, your
main app project must contain Swift code and a bridging header. If your app
project does not contain any Swift code, add a single empty `.swift` file and an
empty bridging header to your app.

If your Xcode project already makes use of Swift code you can safely ignore this
note.

## Usage

> ⚠️ Still using an older version ? Check out the [migration guides](https://github.com/matei-radu/react-native-in-app-browser/wiki/Migration-Guides).

```javascript
import { InAppBrowser } from "@matt-block/react-native-in-app-browser";


// Minimal setup.
InAppBrowser.open("https://www.wikipedia.org/").catch(error => {
  console.log(error);
});


// With platform-specific optional settings.
InAppBrowser.open("https://www.wikipedia.org/", {
  android: {
    //...,
  },
  ios: {
    //...,
  },
}).catch(error => {
  console.log(error);
});


// Using async/await.
async onClickHandler() {
  try {
    await InAppBrowser.open("https://www.wikipedia.org/");
  } catch (error) {
    console.log(error);
  }
}
```

### Startup optimizations (Android only)

Chrome Custom Tabs exposes methods to reduce the start time of a custom tab activity.

For information on how these work, check [Custom Tabs - Example and Usage Optimization](https://chromium.googlesource.com/external/github.com/GoogleChrome/custom-tabs-client/+/08e2c9155aff7296428aae854769c30b4060ae88/README.md#optimization)

```javascript
import { InAppBrowser } from "@matt-block/react-native-in-app-browser";

InAppBrowser.warmup();
InAppBrowser.mayLaunchUrl("https://wikipedia.org");
```

### Programmatically close (iOS only)

It is possible to manually dismiss the iOS in-app instance by calling the method `close`.

```javascript
import { InAppBrowser } from "@matt-block/react-native-in-app-browser";

InAppBrowser.close();
```

This is not possible on Android since Chrome Custom Tabs do not expose such functionality and [Activity workarounds][customtabsmanualclose] would bring this package way out of scope.

## Settings

The main method `open` accepts an object with settings objects for each
platform:

```javascript
import { InAppBrowser } from "@matt-block/react-native-in-app-browser";

InAppBrowser.open("https://www.wikipedia.org/", {
  android: {},
  ios: {}
});
```

Settings can also be initialized separately with `configure` so that each `open` call won't need to specify them each time:

```javascript
import { InAppBrowser } from "@matt-block/react-native-in-app-browser";

// Somewhere in your app initialization logic.
InAppBrowser.configure({
  android: {},
  ios: {}
});

// Other part of your code base.
// Previously initialized settings will apply.
InAppBrowser.open("https://www.wikipedia.org/");
```

Note that `open` will still accept settings as always but will effectively perform a merge between the initialized properties and the provided settings object (which will have priority over the initialized properties):

```javascript
import { InAppBrowser } from "@matt-block/react-native-in-app-browser";

// Somewhere in your app initialization logic.
InAppBrowser.configure({
  android: {
    toolbarColor: "red",
    showTitle: true
  }
});

// Other part of your code base.
// Merged settings for this call will result in:
//
// - toolbarColor: "blue",
// - showTitle: true
InAppBrowser.open("https://www.wikipedia.org/", {
  android: {
    toolbarColor: "blue"
  }
});
```

The properties for each platform settings are described below.

### Android

Each setting is optional. Furthermore, settings with invalid values will be
ignored and their default values will be used.

| Setting               | Type                                                | Default     | Description                                                                                                                |
| --------------------- | --------------------------------------------------- | ----------- | -------------------------------------------------------------------------------------------------------------------------- |
| `toolbarColor`        | `string`                                            | `undefined` | The color to tint the background of the toolbar.<br/>Provided color can be in any [@ctrl/tinycolor][tinycolor] supported format. |
| `showTitle`           | `boolean`                                           | `false`     | Flag to toggle if the page title should be shown in the custom tab.                                                        |
| `closeButtonIcon`     | `string` _(as resolved by importing an image file)_ | `undefined` | Custom close button icon.<br/><br/>Provided icon must be a `.png`, `.jpg`, or `.gif` file.                                 |
| `addDefaultShareMenu` | `boolean`                                           | `false`     | Flag to toggle the default share menu.                                                                                     |

### iOS

Each setting is optional. Furthermore, settings with invalid values will be
ignored and their default values will be used. Also, settings that are not
supported by the iOS version at runtime will be ignored as well.

| Setting                     | Type      | Default     | Description                                                                                                                                                                                    |
| --------------------------- | --------- | ----------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `preferredBarTintColor`     | `string`  | `undefined` | The color to tint the background of the navigation bar and the toolbar.<br/>Provided color can be in any [@ctrl/tinycolor][tinycolor] supported format.<br/><br/>**Available on**: iOS >= 10.0.      |
| `preferredControlTintColor` | `string`  | `undefined` | The color to tint the control buttons on the navigation bar and the toolbar.<br/>Provided color can be in any [@ctrl/tinycolor][tinycolor] supported format.<br/><br/>**Available on**: iOS >= 10.0. |
| `barCollapsingEnabled`      | `boolean` | `true`      | **Available on**: iOS >= 11.0. |
| `dismissButtonStyle`      | `string` | `undefined`      | **Available on**: iOS >= 11.0.   |
| `entersReaderIfAvailable`      | `boolean` | `false`      | A value that specifies whether Safari should enter Reader mode, if it is available.<br/><br/>**Available on**: iOS >= 11.0. |


## Example

An example project showcasing the various configurations can be found in the
`example` directory.
Simply clone this repository, navigate into `example`, install the dependencies
and run the app.

```sh
git clone https://github.com/matei-radu/react-native-in-app-browser.git
cd react-native-in-app-browser/example
yarn install
yarn start
yarn android # or yarn ios
```

## License

Copyright (c) 2018-2020 Matei Bogdan Radu.

This source code is licensed under the MIT license found in the
[LICENSE][license] file in the root directory of this source tree.

<!-- Sources -->

[chromecustomtabs]: https://developer.chrome.com/multidevice/android/customtabs
[customtabsmanualclose]: https://stackoverflow.com/a/41596629/1887860
[sfsafariviewcontroller]: https://developer.apple.com/documentation/safariservices/sfsafariviewcontroller
[license]: https://github.com/matei-radu/react-native-in-app-browser/blob/master/LICENSE
[license_shield]: https://img.shields.io/badge/license-MIT-blue.svg
[prettier_shield]: https://img.shields.io/badge/code_style-prettier-ff69b4.svg
[prettier]: https://github.com/prettier/prettier
[npm]: https://www.npmjs.com/package/@matt-block/react-native-in-app-browser
[npm_shield]: https://img.shields.io/npm/v/@matt-block/react-native-in-app-browser/latest
[npm_2]: https://www.npmjs.com/package/@matt-block/react-native-in-app-browser/v/2.1.3
[npm_2_shield]: https://img.shields.io/npm/v/@matt-block/react-native-in-app-browser/legacy
[npm_1]: https://www.npmjs.com/package/@matt-block/react-native-in-app-browser/v/1.4.1
[npm_1_shield]: https://img.shields.io/badge/npm-v1.4.1-blue
[tinycolor]: https://github.com/TypeCtrl/tinycolor
[react-native-inappbrowser-reborn]: https://github.com/proyecto26/react-native-inappbrowser


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [3.2.2] - 2020-09-30

### Changed
- Bump `@ctrl/tinycolor` to version `3.1.2`.
- Bump other devDependencies.

## [3.2.1] - 2020-05-17

### Changed
- Bump Kotlin to version `1.3.72`.
- Bump most JS/TS dependencies.

## [3.2.0] - 2020-02-29

### Added

- iOS specific: `entersReaderIfAvailable` setting to instruct Safari to enter in Reader mode, if available (thanks to [@fonov](https://github.com/fonov)).
- iOS specific: `dismissButtonStyle` setting to change the label of the dismiss button.

## [3.1.4] - 2020-01-12

### Changed

- `tinycolor2` was replaced with the fork `@ctrl/tinycolor` because the the former has not received any update since 2016.

## [3.1.3] - 2019-12-14

### Fixed

- The React peer dependency is now `16.8.6` as needed by RN 0.60.0 and upwards.

### Changed

- Bump TypeScript to version `3.7.3`.

## [3.1.2] - 2019-10-19

### Added

- Compatibility with React Native `0.61.0`.
- Flow typings for Android `warmup` and `mayLaunchUrl` methods.

### Fixed

- On iOS, `threading violation` exception thrown while invoking `closeInApp` (thanks to [@fonov](https://github.com/fonov)).

## [3.1.1] - 2019-09-01

### Fixed

- On Android, crash on app launch when there are no Custom Tabs compatible browsers available.

## [3.1.0] - 2019-08-06

### Added

- Android specific `warmup` and `mayLaunchUrl` methods to reduce custom tab activity start time.

## [3.0.0] - 2019-07-13

### Added

- Compatibility with React Native `0.60.0` and, with it, [autolinking](https://github.com/react-native-community/cli/blob/master/docs/autolinking.md): no need to manually link anymore.

### Removed

- Compatibility with React Native versions below `0.60.0`. If you need to use an older version of React Native, check the [compatibility table](https://github.com/matei-radu/react-native-in-app-browser#compatibility) to use the appropriate package.

### Changed

- iOS podspec file has been updated to support React Native `0.60.0`.
- Bump Kotlin to version `1.3.31`.
- Remove explicit Android Build Tools version from Gradle configuration.

## [2.0.0] - 2019-06-01

### Removed

The following deprecated functions have been removed. Please see
[Migrating from 1.x to 2.x](https://github.com/matt-block/react-native-in-app-browser/wiki/Migrating-from-1.x-to-2.x) for more details.

- `openInApp`, use `InAppBrowser.open` instead.
- `initialize`, use `InAppBrowser.configure` instead.
- `closeInAppInstance`, use `InAppBrowser.close` instead.

## [1.4.1] - 2019-04-19

### Fixed

- Packaging script typo that prevended NPM interoperability (thanks to [@a6kme](https://github.com/a6kme)).

### Changed

- Android `getBitmapFromUriOrDrawable` has been refactored for better maintainability.
- Android `convertDpToPixel` has been replaced with Facebook's `PixelUtil.toPixelFromDIP` since it's more tested and reliable.

## [1.4.0] - 2019-03-01

### Added

- Object `InAppBrowser` with methods:
  - `open` as a drop-in replacement for `openInApp`,
  - `configure` as a drop-in replacement for `initialize`,
  - `close` as a drop-in replacement for `closeInAppInstance`.

### Deprecated

The following functions have been deprecated. Please see
[Migrating from 1.x to 2.x](https://github.com/matt-block/react-native-in-app-browser/wiki/Migrating-from-1.x-to-2.x) for more details.

- `openInApp`, use `InAppBrowser.open` instead.
- `initialize`, use `InAppBrowser.configure` instead.
- `closeInAppInstance`, use `InAppBrowser.close` instead.

## [1.3.0] - 2019-02-27

### Added

- Example project in the `example` directory.

### Changed

- Color properties (`toolbarColor`, `preferredBarTintColor` and `preferredControlTintColor`) now
  accept more color formats. Please refer to TinyColor's [Accepted String Input](https://github.com/bgrins/TinyColor#accepted-string-input) to see all available formats.
- The iOS `.podspec` will now automatically stay in sync with `package.json` (thanks to [@ericlewis](https://github.com/ericlewis)).

## [1.2.7] - 2019-02-21

### Added

- Compatibility with React Native `0.59.0`.

### Fixed

- Missing Flow definitions for `closeInAppInstance`.
- Outdated iOS podspec file.

### Changed

- Bump Kotlin to version `1.3.21`.
- Minor refactoring of `settings.ts`.

## [1.2.4] - 2019-01-29

### Added

- Compatibility with React Native `0.58.0`.

### Fixed

- Possible iOS crashes due to thread clashing (thanks to [@yoneapp](https://github.com/yoneapp))

### Changed

- Bump Android target from `26` to `27` (same as React Native `0.58.0`).
- Bump Android build tools from `27.0.3` to `28.0.2` (same as React Native `0.58.0`).
- Bump Android compile vesion from `27` to `28` (same as React Native `0.58.0`).
- Bump Android Gradle tools from `3.1.4` to `3.2.0` (same as React Native `0.58.0`).

## [1.2.1] - 2019-01-10

### Fixed

- Typo in the iOS `openInApp` function that prevented compilation.

## [1.2.0] - 2019-01-08

### Added

- `closeInAppInstance`: programmatically close the in-app browser (works for iOS only).

## [1.1.2] - 2018-12-05

### Changed

- Relax Xcode project compatibility version to `8.0`.

## [1.1.1] - 2018-10-30

### Changed

- Bump Kotlin to version `1.3.0`.
- Update other dependencies.

## [1.1.0] - 2018-10-14

### Added

- `initialize`: initialize the settings of the in-app browser so that each `openInApp` call won't need to specify them each time.

### Changed

- Minor refactoring.

## [1.0.1] - 2018-10-07

### Changed

- Test release necessary to evaluate the CircleCI integration and auto publishing to NPM.

## [1.0.0] - 2018-10-03

### Added

- `openInApp`: open a valid http(s) URL with an in-app browser.

[Unreleased]: https://github.com/matei-radu/react-native-in-app-browser/compare/v3.2.2...HEAD

src/settings.ts
/**
 * Copyright (c) 2018-2020, Matei Bogdan Radu <opensource@mateiradu.dev>
 *
 * This source code is licensed under the MIT license found in the LICENSE
 * file in the root directory of this source tree.
 */

import { PlatformOSType, Image } from 'react-native';
import { TinyColor, ColorInput } from '@ctrl/tinycolor';

export interface Settings {
  /**
   * Android specific settings.
   *
   * All settings are optional and invalid ones will be ignored.
   *
   * The implementation uses Chrome Custom Tabs.
   */
  android?: SettingsAndroid;

  /**
   * iOS specific settings.
   *
   * All settings are optional and invalid ones will be ignored.
   *
   * The implementation uses Safari View Manager.
   */
  ios?: SettingsIOS;
}

export interface SettingsAndroid {
  /**
   * The color to tint the background of the toolbar.
   *
   * **Note**: if the color string is invalid, this setting will be ignored.
   */
  toolbarColor?: ColorInput;

  /**
   * Flag to toggle if the title should be shown in the custom tab.
   *
   * **Note**: if the value is invalid, this setting will be ignored.
   */
  showTitle?: boolean;

  /**
   * Custom close button icon.
   *
   * Provided icon must be a `.png`, `.jpg`, or `.gif` file.
   *
   * **Note**: if icon asset is invalid, this setting will be ignored.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  closeButtonIcon?: any;

  /**
   * Flag to toggle the default share menu.
   *
   * **Note**: if the value is invalid, this setting will be ignored.
   */
  addDefaultShareMenu?: boolean;
}

export interface SettingsIOS {
  /**
   * The color to tint the background of the navigation bar and the toolbar.
   *
   * **Available on**: iOS >= 10.0.
   *
   * **Note**: if the color string is invalid or if the current iOS version
   * is < 10.0, this setting will be ignored.
   */
  preferredBarTintColor?: ColorInput;

  /**
   * The color to tint the control buttons on the navigation bar and the
   * toolbar.
   *
   * **Available on**: iOS >= 10.0.
   *
   * **Note**: if the color string is invalid or if the current iOS version
   * is < 10.0, this setting will be ignored.
   */
  preferredControlTintColor?: ColorInput;

  /**
   * **Available on**: iOS >= 11.0.
   *
   * **Note**: if the value is invalid or if the current iOS version
   * is < 11.0, this setting will be ignored.
   */
  barCollapsingEnabled?: boolean;

  /**
   * A value that specifies whether Safari should enter Reader mode, if it is available.
   *
   * **Available on**: iOS >= 11.0.
   *
   * **Note**: if the value is invalid or if the current iOS version
   * is < 11.0, this setting will be ignored.
   */
  entersReaderIfAvailable?: boolean;

  /**
   * **Available on**: iOS >= 11.0.
   *
   * **Note**: if the value is invalid or if the current iOS version
   * is < 11.0, this setting will be ignored.
   */
  dismissButtonStyle?: IOSDismissButtonStyle;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any;
}

const iosDismissButtonStyles = ['done', 'close', 'cancel'] as const;
type IOSDismissButtonStyle = typeof iosDismissButtonStyles[number]

/**
 * Default settings.
 *
 * These values can be augmented throgh initialization.
 */
export const defaultSettings: Required<Settings> = {
  android: {},
  ios: {},
};

function sanitizeAndroid(settings?: SettingsAndroid): SettingsAndroid {
  const sanitized = { ...defaultSettings.android };

  if (!settings) {
    return sanitized;
  }

  const toolbarColor = new TinyColor(settings.toolbarColor);
  if (toolbarColor.isValid) {
    sanitized.toolbarColor = toolbarColor.toHexString();
  }

  if (typeof settings.showTitle === 'boolean') {
    sanitized.showTitle = settings.showTitle;
  }

  if (settings.closeButtonIcon) {
    sanitized.closeButtonIcon = Image.resolveAssetSource(
      settings.closeButtonIcon
    ).uri;
  }

  if (typeof settings.addDefaultShareMenu === 'boolean') {
    sanitized.addDefaultShareMenu = settings.addDefaultShareMenu;
  }

  return sanitized;
}

function sanitizeIOS(settings?: SettingsIOS): SettingsIOS {
  const sanitized = { ...defaultSettings.ios };

  if (!settings) {
    return sanitized;
  }

  const colors = ['preferredBarTintColor', 'preferredControlTintColor'];
  colors.forEach(color => {
    const parsedColor = new TinyColor(settings[color]);
    if (parsedColor.isValid) {
      sanitized[color] = parsedColor.toHexString();
    }
  });

  if (typeof settings.barCollapsingEnabled === 'boolean') {
    sanitized.barCollapsingEnabled = settings.barCollapsingEnabled;
  }

  if (typeof settings.entersReaderIfAvailable === 'boolean') {
    sanitized.entersReaderIfAvailable = settings.entersReaderIfAvailable;
  }

  if (settings.dismissButtonStyle && iosDismissButtonStyles.includes(settings.dismissButtonStyle)) {
    sanitized.dismissButtonStyle = settings.dismissButtonStyle;
  }

  return sanitized;
}

/**
 * Sanitize the settings based on the running OS.
 *
 * Provided settings will be merged with the default ones.
 * Also, in case of same properties, provided ones have priority
 * over defaults.
 */
export function sanitize(os: 'android', settings?: Settings): SettingsAndroid;
export function sanitize(os: 'ios', settings?: Settings): SettingsIOS;
export function sanitize(os: PlatformOSType, settings?: Settings): {};
export function sanitize(
  os: PlatformOSType,
  settings?: Settings,
): SettingsAndroid | SettingsIOS {
  switch (os) {
    case 'android':
      return sanitizeAndroid(settings?.android);
    case 'ios':
      return sanitizeIOS(settings?.ios);
    // Other platforms in the future.
    default:
      return {};
  }
}

/**
 * Initializes the platform-specific settings for the in-app browser
 * experience.
 *
 * This utility function is useful when `openInApp` is used in several
 * portions of the application code base as it allows to provide the
 * settings only once instead of specifing them with each call.
 */
export function initialize(settings: Settings): void {
  // First, reset directly as `sanitize` will otherwise merge with
  // previous defaults: it would not be possible to remove properties.
  defaultSettings.android = {};
  defaultSettings.ios = {};

  defaultSettings.android = sanitize('android', settings);
  defaultSettings.ios = sanitize('ios', settings);
}
